cli/azd/pkg/commands/pipeline/azdo_provider.go

```diff
@@ -7,65 +7,528 @@ import (
-func (p *AzdoHubScmProvider) preConfigureCheck(ctx context.Context, console input.Console) error {
-	return errors.New("not implemented")
+func (p *AzdoScmProvider) preConfigureCheck(ctx context.Context, console input.Console) error {
+	_, err := azdo.EnsurePatExists(ctx, p.Env, console)
+	if err != nil {
+		return err
+	}
+
+	_, err = azdo.EnsureOrgNameExists(ctx, p.Env, console)
+	return err
+}
+
+// helper function to save configuration values to .env file
+func (p *AzdoScmProvider) saveEnvironmentConfig(key string, value string) error {
+	p.Env.Values[key] = value
+	err := p.Env.Save()
+	if err != nil {
+		return err
+	}
+	return nil
 }
 
 // name returns the name of the provider
-func (p *AzdoHubScmProvider) name() string {
+func (p *AzdoScmProvider) name() string {
 	return "Azure DevOps"
 }
 
 // ***  scmProvider implementation ******
 
+// stores repo details in state for use in other functions. Also saves AzDo project details to .env
+func (p *AzdoScmProvider) StoreRepoDetails(ctx context.Context, repo *azdoGit.GitRepository) error {
+	repoDetails := p.getRepoDetails()
+	repoDetails.repoName = *repo.Name
+	repoDetails.remoteUrl = *repo.RemoteUrl
+	repoDetails.repoWebUrl = *repo.WebUrl
+	repoDetails.sshUrl = *repo.SshUrl
+	repoDetails.repoId = repo.Id.String()
+
+	err := p.saveEnvironmentConfig(azdo.AzDoEnvironmentRepoIdName, p.repoDetails.repoId)
+	if err != nil {
+		return fmt.Errorf("error saving repo id to environment %w", err)
+	}
+
+	err = p.saveEnvironmentConfig(azdo.AzDoEnvironmentRepoName, p.repoDetails.repoName)
+	if err != nil {
+		return fmt.Errorf("error saving repo name to environment %w", err)
+	}
+
+	err = p.saveEnvironmentConfig(azdo.AzDoEnvironmentRepoWebUrl, p.repoDetails.repoWebUrl)
+	if err != nil {
+		return fmt.Errorf("error saving repo web url to environment %w", err)
+	}
+
+	return nil
+}
+
+// prompts the user for a new AzDo Git repo and creates the repo
+func (p *AzdoScmProvider) createNewGitRepositoryFromInput(ctx context.Context, console input.Console) (string, error) {
+	connection, err := p.getAzdoConnection(ctx)
+	if err != nil {
+		return "", err
+	}
+
+	var repo *azdoGit.GitRepository
+	for {
+		name, err := console.Prompt(ctx, input.ConsoleOptions{
+			Message:      "Enter the name for your new Azure Devops Repository OR Hit enter to use this name:",
+			DefaultValue: p.repoDetails.projectName,
+		})
+		if err != nil {
+			return "", fmt.Errorf("asking for new project name: %w", err)
+		}
+
+		var message string
+		newRepo, err := azdo.CreateRepository(ctx, p.repoDetails.projectId, name, connection)
+		if err != nil {
+			message = err.Error()
+		}
+		if strings.Contains(message, fmt.Sprintf("A Git repository with the name %s already exists.", name)) {
+			console.Message(ctx, fmt.Sprintf("error: the repo name '%s' is already in use\n", name))
+			continue // try again
+		} else if strings.Contains(message, "TF401025: 'repoName' is not a valid name for a Git repository.") {
```

Comment:
Just checking here - the `'repoName'` part of this error message never changes based on what name you try to use, right?